service/dap/server.go

```diff
@@ -469,6 +469,123 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+
+		// Validate rr availability
+		if err := gdbserial.CheckRRAvailable(); err != nil {
+			s.sendErrorResponseWithURL(request.Request, FailedToInitialize, "Unable to find  rr in record configuration", err.Error(), "https://github.com/rr-debugger/rr/wiki/Building-And-Installing", "rr installation page")
+			return
+		}
+
+		// Record only works on rr trace directories
+		traceDirectory, ok := request.Arguments["traceDirectory"].(string)
+		if !ok || traceDirectory == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The traceDirectory attribute is missing in record configuration.")
+			return
+		}
+
+		buildFlags := ""
+		buildFlagsArg, ok := request.Arguments["buildFlags"]
+		if ok {
+			buildFlags, ok = buildFlagsArg.(string)
+			if !ok {
+				s.sendErrorResponse(request.Request,
+					FailedToLaunch, "Failed to launch",
+					fmt.Sprintf("'buildFlags' attribute " +
+						"%v' in record configuration is not a string.", buildFlagsArg))
+				return
+			}
+		}
+
+		output, ok := request.Arguments["output"].(string)
+		if !ok || output == "" {
+			output = defaultDebugBinary
+		}
+
+		debugname, err := filepath.Abs(output)
+		err = gobuild.GoBuild(debugname, []string{program}, buildFlags)
+		if err != nil {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to build recording target",
+				fmt.Sprintf("Build error: %s", err.Error()))
+			return
+		}
+
+		traceDir, err := gdbserial.Record([]string{program}, filepath.Dir(program), true, s.config.Debugger.Redirects)
+		if err != nil {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to record trace from target",
+				fmt.Sprintf("Recording error: %s", err.Error()))
+			return
+		}
+		if traceDir == "" {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to record trace from target",
+				"No trace result found")
+			return
+		}
+	}
+
+	if mode == "replay" {
+		// Check against replay parameters and decide the backend based on which one was received
+		coreDumpPath, coreDumpPathOk := request.Arguments["coreDumpPath"].(string)
+		traceDirectory, traceDirectoryOk := request.Arguments["traceDirectory"].(string)
+
+		coreDumpPathExists := coreDumpPathOk && (coreDumpPath != "")
+		traceDirectoryExists := traceDirectoryOk && (traceDirectory != "")
+		if coreDumpPathExists && traceDirectoryExists {
+			s.sendErrorResponse(request.Request,
+				FailedToLaunch, "Failed to launch",
+				"The replay configuration must have either a coreDumpPath or a traceDirectory. Both attributes were found.")
+			return
+		}
+
+		//TODO(lgomez): define default backend to use for this scenario as 'core' itself is only an internal value in order to differentiate from rr
+		backend := "core"
+		if traceDirectoryExists {
+			backend = "rr"
+			// Enable StepBack controls
+			s.send(&dap.CapabilitiesEvent{ Event: *newEvent("capabilities"), Body: dap.CapabilitiesEventBody{Capabilities: dap.Capabilities{ SupportsStepBack: true }}})
```

Comment:
This should be moved after the `debugger.New` call, if it succeeds check what the value of `s.config.Debugger.Backend` is and if it is `rr` send this event.